### DISCUSSION TITLE
Check if a key exists in a dictionary

### DISCUSSION THREAD

**Discussion by lucasbracher** [zeroSteiner/rule-engine#87] [General]
Hello! I'm trying to figure out a way to check if a key exists in a dictionary. How can I do that, please?
Thank you!

**Comment by zeroSteiner** [comment]
You use the `in` keyword, just like in Python.

```
PYTHONPATH=$(pwd)/lib python -m rule_engine.debug_repl --edit-console
edit the 'context' and 'thing' objects as necessary
>>> thing = {'person': {'name': 'Spencer'}}
>>> 
exiting the edit console...
rule > 'name' in person
result: 
True
rule > 'age' in person
result: 
False
rule > 
```

**Reply by lucasbracher** [reply]
Thank you! Just another question, in this case, how could I check if "person" is in the outer dictionary?

**Reply by zeroSteiner** [reply]
So when the rule is evaluated, the `thing` defines all the variables that are within the context. You can't easily check if a variable is defined. I would recommend normalizing your object before running the evaluation to initialize missing values to a sane default.

If you're using the default configuration (as I was in my example) where `thing` is a dictionary that responds to `getitem`, then you'd just define another dictionary with defaults and merge them like `{'balance': 0, 'email': None} | person`:

```
In [1]: import rule_engine

In [2]: rule = rule_engine.Rule('balance > 0')

In [3]: defaults = {'balance': 0, 'email': None}

In [4]: person = {'name': 'Spencer'}

In [5]: rule.evaluate(defaults | person)
Out[5]: False

In [6]: person['balance'] = 100

In [7]: rule.evaluate(defaults | person)
Out[7]: True 
```

This approach is more precise and allows the default values to be of different data types. That is to say balance is initially zero and email defaults to null. Alternatively, you can play it fast and loose, by setting the default_value in the context to None which will cause any variable that is not defined to be None, or really whatever else you chose like zero an empty array, whatever. There's more information in the docs if you want to take that approach https://zerosteiner.github.io/rule-engine/getting_started.html#setting-a-default-value
